pkg/cri/resource-manager/resource-manager.go

```diff
@@ -154,33 +159,66 @@ func (m *resmgr) Stop() {
 	m.Lock()
 	defer m.Unlock()
 
+	if m.signals != nil {
+		close(m.signals)
+		m.signals = nil
+	}
+
 	m.configServer.Stop()
 	m.relay.Stop()
 	m.stopEventProcessing()
 }
 
 // SetConfig pushes new configuration to the resource manager.
 func (m *resmgr) SetConfig(conf *config.RawConfig) error {
-	m.Info("applying new configuration...")
+	m.Info("applying new configuration from agent...")
 
 	m.Lock()
 	defer m.Unlock()
+	return m.setConfig(conf)
+}
 
-	// TODO: save current configuration and roll back if some controllers fail to start
+// setConfigFromFile pushes new configuration to the resource manager from a file.
+func (m *resmgr) setConfigFromFile(path string) error {
+	m.Info("applying new configuration from file %s...", path)
+
+	m.Lock()
+	defer m.Unlock()
+	return m.setConfig(path)
+}
 
-	if err := pkgcfg.SetConfig(conf.Data); err != nil {
-		m.Error("new configuration was rejected: %v", err)
+// setConfig pushes new configuration from the agent or a file.
+func (m *resmgr) setConfig(src interface{}) error {
+	var pending *config.RawConfig
+	var err error
+
+	switch cfg := src.(type) {
+	case *config.RawConfig:
+		pending = cfg
+		err = pkgcfg.SetConfig(cfg.Data)
+	case string:
+		err = pkgcfg.SetConfigFromFile(cfg)
```

Comment:
I'd somehow document that it's intentional that we're not caching config that is read from a file. It's not outright obvious and at least I tend to forget it

Reply:
Added comments about never storing forced configuration in the cache. One in the switch pending is not set for forced configuration, and another where a pending config is stored.